actions/stale
# Add support for adding close label

Closes https://github.com/actions/stale/issues/134.

### Description 

This pull request is introducing two new params to the public interface:

```yaml
close-issue-label:
    description: 'The label to apply when an issue is closed.'

close-pr-label:
    description: 'The label to apply when a pull request is closed.'
```

These are new labels, that can be applied when an issue / a pr is closed.

### Testing 

**Unit Tests**

Unfortunately I was not able to implement unit tests for this functionality due to the current limitations of `IssueProcessor` verification - we cannot verify how an issue / a pr has been affected by params like `close-issue-message` or `stale-issue-label`.

We could create a separate issue / pr to improve testing part. One of the idea could be using [nock](https://github.com/nock/nock) to mock the calls to the GitHub API. That would allow us for more complex verification of the modifications:

```typescript
test('processing a stale issue will close it with close label when close-issue-label is set', async () => {
  const TestIssueList: Issue[] = [
    generateIssue(
      1,
      'A stale issue that should be closed',
      '2020-01-01T17:00:00Z',
      false,
      ['Stale']
    )
  ];

  const opts = {...DefaultProcessorOptions};
  opts.closeIssueLabel = 'Closed'

  const processor = new IssueProcessor(
    DefaultProcessorOptions,
    async p => (p == 1 ? TestIssueList : []),
    async (num, dt) => [],
    async (issue, label) => new Date().toDateString()
  );

  // 🔨  GitHub API mocking with nock 

  // process our fake issue list
  await processor.processIssues(1);

  // 🔨  verify if the label `Closed` is added to the issue
});
```

Let me please know what you think 😊 

**Manual testing**

```yaml
name: Closing stale issues
on:
  schedule:
  - cron: "0 0 * * *"

jobs:
  stale:
    runs-on: ubuntu-latest
    steps:
    - uses: mollyIV/stale@feature/close-label
      with:
        repo-token: ${{ secrets.BOT_GITHUB_TOKEN }}
        days-before-stale: 1
        days-before-close: 1
        stale-issue-message: > 
          There hasn't been any activity on this issue recently. Due to the high number of incoming GitHub notifications, we have to clean some of the old issues, as many of them have already been resolved with the latest updates.
          
          Please make sure to update to the latest `fastlane` version and check if that solves the issue. Let us know if that works for you by adding a comment :+1:
          
          Friendly reminder: contributions are always welcome! Check out [CONTRIBUTING.md](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md) for more information on how to help with `fastlane` and feel free to tackle this issue yourself :muscle:
        stale-issue-label: 'status: waiting-for-reply'
        close-issue-message: >
          This issue will be auto-closed because there hasn't been any activity for a few months. Feel free to [open a new one](https://github.com/fastlane/fastlane/issues/new) if you still experience this problem 👍
        stale-pr-label: 'status: needs-attention'
        skip-stale-pr-message: true
        stale-pr-message: 'Stale'
        close-pr-label: 'issue-status: auto-closed'
        close-issue-label: 'pr-status: auto-closed'
```

<img width="1251" alt="Screenshot 2020-08-11 at 08 28 10" src="https://user-images.githubusercontent.com/10795657/89864692-ad812000-dbac-11ea-974d-03dd7436b96f.png">

<img width="1246" alt="Screenshot 2020-08-11 at 08 27 53" src="https://user-images.githubusercontent.com/10795657/89864712-b540c480-dbac-11ea-90c1-45de6cc45f85.png">

> I have used my personal token, so the comments come from my user. 

🎊
